cdap-messaging-ext-spanner/src/main/java/io/cdap/cdap/messaging/spanner/SpannerMessagingService.java

```diff
@@ -293,18 +311,37 @@ public RollbackDetail publish(StoreRequest request)
     if (!batch.isEmpty()) {
       int sequenceId = 0;
       List<Mutation> batchCopy = new ArrayList<>(batch.size());
+      long estimatedBatchSize = 0;
       // We need to batch less than fetch limit since we read for publish_ts >= last_message.publish_ts
       // see fetch for explanation of why we read for publish_ts >= last_message.publish_ts and
       // not publish_ts > last_message.publish_ts
       while (!batch.isEmpty()) {
         StoreRequest headRequest = batch.poll();
         for (byte[] payload : headRequest) {
-          //TODO: [CDAP-21094] Breakdown messages with larger payload into parts.
-          Mutation mutation = Mutation.newInsertBuilder(getTableName(headRequest.getTopicId()))
-              .set(SEQUENCE_ID_FIELD).to(sequenceId++).set(PAYLOAD_SEQUENCE_ID_FIELD).to(0)
-              .set(PUBLISH_TS_FIELD).to("spanner.commit_timestamp()").set(PAYLOAD_FIELD)
-              .to(ByteArray.copyFrom(payload)).build();
-          batchCopy.add(mutation);
+          int payloadSequenceId = 0;
+          List<Mutation> mutationsForCurrentPayload = new ArrayList<>();
+          int remainingChunksCount = (payload.length - 1) / MAX_PAYLOAD_SIZE_IN_BYTES;
+          int length = payload.length;
+          int offset = 0;
+          while (offset < length) {
+            int chunkSize = Math.min(MAX_PAYLOAD_SIZE_IN_BYTES, length - offset);
+            byte[] payloadChunk = Arrays.copyOfRange(payload, offset, offset + chunkSize);
```

Comment:
I believe we can optimize here by avoiding unnecessary array copy in 99% of cases by setting payloadChunk to payload if remainingChunksCount=0. Right?